x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java

```diff
@@ -212,6 +238,29 @@ private static Failure fail(Node<?> source, String message, Object... args) {
                 failures.addAll(localFailures);
             });
         }
+        
+        // gather metrics
+        if (failures.isEmpty() && metrics != null) {
```

Comment:
To avoid incrementing the counter multiple times you have several options:
1. use `plan/expression.anyMatch()` for each check, this will result in multiple iterations though, one per check
2. use a method variable of a Map of booleans. This is essentially a `BitSet` where the ordinal of the enum turns on the feature is found.
After traversing the tree, traverse the `BitSet` and set the appropriate metric:
```java
BitSet b = new BitSet(Features.values().size());
...

if (p instanceof Aggregate) {
        b.set(GROUPBY.ordinal);
}
...

}
```

Depending on the carrotsearch or java.util BitSet you can have slightly different ways of iterating but essentially you find the index of the next set bit, convert that into an ordinal, get the enum and inc the metric.

```java
for( i = nextBit() ; i !=0; i = nextBit(i)) {
   metric.inc(Metrics.values()[i]);
}
```

Reply:
@costin thank you for your comments and valuable suggestions. I think both options can be used at the same time:
- `anyMatch()` to stop after the first match is found
- `BitSet` to replace the `Map<FeatureSet, Boolean>` current implementation in `Metrics`

But I cannot do the BitSet initialization in the Verifier, because the possible multiple increments happen because `Verifier.verify(LogicalPlan)` is called multiple times in the analysis process. So, the BitSet "control" should be outside the `Verifier`.